File: contrail/src/contrail/Item.go
package contrail

import (
	"math/big"
	"time"
	"contrail/id"
)

type PropertyChangeListener func(source Item, propertyName string, oldValue interface{}, newValue interface{}) 


/**
 * An interface that every object to be stored in a Contrail database is required to implement.
 * An Item is a collection of property values indexed by property names.
 * All items have an identifier that denotes a unique path to the item's 
 * location in the database.
 * An Item is just a collection of properties.
 * Property values must be one of the following:
 * 	string,
 *  time.Time, 
 * 	int64, float64, byte, math.Int, math.Rat
 * 	bool,
 * 	contrail.id.Identifier,
 * 	contrail.Item,
 *  Go Slice, where items in slice are other valid property values
 *
 *  Item has methods for tracking changes to properties.
 *  @see ***PropertyChangeListener
 * 
 *  Items are thread-safe.
 * 
 * @author Ted Stockwell
 */
type Item interface {
	/**
	 * @return a unique identifier used to locate an IEntity in storage.
	 */
	Identifier() *id.Identifier 
	Copy() Item 
	
	GetString(propertyName string) string 
	GetIdentifier(propertyName string) *id.Identifier 
	GetRat(propertyName string) big.Rat 
	GetInt(propertyName string) big.Int 
	GetBool(propertyName string) bool 
	GetItem(propertyName string) Item 
	GetByte(propertyName string) byte 
	GetTime(propertyName string) time.Time 
	GetFloat(propertyName string) float64 
	GetLong(propertyName string) int64 
	GetSlice(propertyName string) []interface{} 
	
	AllProperties() map[string]interface{}
	IndexedProperties() map[string]interface{}
	UnindexedProperties() map[string]interface{}

	GetProperty(propertyName string) interface{}
	HasProperty(propertyName string) bool
	RemoveProperty(propertyName string) Item
	SetProperties(map[string]interface{})
	SetProperty(propertyName string, value interface{})
	SetUnindexedProperty(propertyName string, value interface{})

    /**
     * Add a PropertyChangeListener to the listener list.
     * The listener is registered for all properties.
     * The same listener object may be added more than once, and will be called
     * as many times as it is added.
     * If <code>listener</code> is null, no exception is thrown and no action
     * is taken.
     *
     * @param listener  The PropertyChangeListener to be added
     */
    AddPropertyChangeListener(listener PropertyChangeListener)
    
    /**
     * Remove a PropertyChangeListener from the listener list.
     * This removes a PropertyChangeListener that was registered
     * for all properties.
     * If <code>listener</code> was added more than once to the same event
     * source, it will be notified one less time after being removed.
     * If <code>listener</code> is null, or was never added, no exception is
     * thrown and no action is taken.
     *
     * @param listener  The PropertyChangeListener to be removed
     */
    RemovePropertyChangeListener(listener PropertyChangeListener)
    

    /**
     * Add a PropertyChangeListener for a specific property.  The listener
     * will be invoked only when a call on firePropertyChange names that
     * specific property.
     * The same listener object may be added more than once.  For each
     * property,  the listener will be invoked the number of times it was added
     * for that property.
     * If <code>propertyName</code> or <code>listener</code> is null, no
     * exception is thrown and no action is taken.
     *
     * @param propertyName  The name of the property to listen on.
     * @param listener  The PropertyChangeListener to be added
     */

    AddPropertyChangeListener(propertyName string, listener PropertyChangeListener)

    /**
     * Remove a PropertyChangeListener for a specific property.
     *
     * @param propertyName  The name of the property that was listened on.
     * @param listener  The PropertyChangeListener to be removed
     */
    RemovePropertyChangeListener(propertyName string, listener PropertyChangeListener)
	
}

File: ContrailNet/ContrailNet/Contrail/Item.cs
using System;

namespace Contrail {


/**
 * An object that can be stored in a Contrail database.
 * All items have an identifier that denotes a unique path to the item's 
 * location in the database.
 * An Item is just a collection of properties.
 * Property values must be one of the following:
 * 	String,
 *  Date, 
 * 	Integer, Byte, Long, Float, Double, BigInteger, or BigDecimal
 * 	Boolean,
 * 	Identifier,
 * 	Item,
 *  List or Set, where items in collections are valid property values
 *  Map, where keys are valid scalar property values 
 *  	(that is, keys may NOT be Item, List, Set, or Map) 
 *  	and values are valid property values     
 * 
 * @author Ted Stockwell
 */
public class Item : Entity, ICloneable
{
	
	/**
	 * A reserved property name used to refer to the ID of an item. 
	 */
	public static readonly String KEY_ID= "__ID__";
	/**
	 * A reserved property name used to refer to the ID of an item. 
	 */
	public static readonly String KEY_KIND = "__KIND__";
	
	private TreeMap<String, Object> _indexedProperties= new TreeMap<String, Object>(); 
	private TreeMap<String, Object> _unindexedProperties= new TreeMap<String, Object>(); 
	
	
	public Item(Identifier path) {
		super(path);
		setProperty(KEY_KIND, getClass().getName());
	}
	
	public Item(String path) {
		this(Identifier.create(path));
	}
	protected Item() { }
	
	public Item(Item item) {
		super(item.getId());
		for (Map.Entry<String, Object> entry:item._indexedProperties.entrySet()) 
			_indexedProperties.put(entry.getKey(), entry.getValue());
		for (Map.Entry<String, Object> entry:item._unindexedProperties.entrySet()) 
			_unindexedProperties.put(entry.getKey(), entry.getValue());
	}
	public Item(Identifier parent, String child) {
		this(Identifier.create(parent, child));
	}
	
	public BigDecimal getBigDecimal(String propertyName) {
		return ConversionUtils.toBigDecimal(getProperty(propertyName));
	}

	public BigInteger getBigInteger(String propertyName) {
		return ConversionUtils.toBigInteger(getProperty(propertyName));
	}

	public bool getBoolean(String propertyName) {
		return ConversionUtils.toBoolean(getProperty(propertyName));
	}

	public Item getItem(String propertyName) {
		return (Item)_indexedProperties.get(propertyName);
	}

	public byte getByte(String propertyName) {
		return ConversionUtils.toByte(getProperty(propertyName));
	}

	public Date getDate(String propertyName) {
		return ConversionUtils.toDate(getProperty(propertyName));
	}

	public double getDouble(String propertyName) {
		return ConversionUtils.toDouble(getProperty(propertyName));
	}

	public float getFloat(String propertyName) {
		return ConversionUtils.toFloat(getProperty(propertyName));
	}

	public int getInt(String propertyName) {
		return ConversionUtils.toInteger(getProperty(propertyName));
	}

	public List<T> getList<T>(String propertyName) {
		return (List<T>)getProperty(propertyName);
	}

	public long getLong(String propertyName) {
		return ConversionUtils.toLong(getProperty(propertyName));
	}

	public Map<K, V> getMap<K, V>(String propertyName) {
		return (Map<K, V>)getProperty(propertyName);
	}

	public Map<String, Object> getAllProperties() {
		TreeMap<String, Object> map= new TreeMap<String, Object>(_indexedProperties);
		map.putAll(_unindexedProperties);
		return map;
	}

	public Map<String, Object> getUnindexedProperties() {
		return Collections.unmodifiableMap(_unindexedProperties);
	}

	public Map<String, Object> getIndexedProperties() {
		return Collections.unmodifiableMap(_indexedProperties);
	}

	public T getProperty<T>(String propertyName) {
		T t= (T) _indexedProperties.get(propertyName);
		if (t == null)
			t= (T) _unindexedProperties.get(propertyName);
		return t;
	}

	public Set<T> getSet<T>(String propertyName) {
		return (Set<T>)getProperty(propertyName);
	}

	public String getString(String propertyName) {
		return ConversionUtils.toString(getProperty(propertyName));
	}

	public boolean hasProperty(String propertyName) {
		if (_indexedProperties.containsKey(propertyName))
			return true;
		return _unindexedProperties.containsKey(propertyName);
	}

	public Item removeProperty(String propertyName) {
		_indexedProperties.remove(propertyName);
		_unindexedProperties.remove(propertyName);
		return this;
	}

	public Item setPropertiesFrom(Map<String, Object> src) {
		for (Map.Entry<String, Object> entry:src.entrySet()) {
			setProperty(entry.getKey(), entry.getValue());
		}
		return this;
	}
	
	public Item setProperty(String propertyName, Object value) {
		_indexedProperties.put(propertyName, value);
		_unindexedProperties.remove(propertyName);
		return this;
	}
	
	public Item setUnindexedProperty(String propertyName, Object value) {
		_unindexedProperties.put(propertyName, value);
		_indexedProperties.remove(propertyName);
		return this;
	}

	override public Item Clone()  {
		return new Item(this);
	}

}

}